### PR TITLE
[ADD] Changeable optimizer

### PR DESCRIPTION
**Pull request by PPierzc**
Added the option to select the optimizer.

### CODE REVIEW

File: fitgabor/trainer.py [hunk 1]
```diff
@@ -3,32 +3,39 @@
...
             gabor_std_constrained = fixed_std * gabor / gabor_std
-        
+
         loss = -model_neuron(gabor_std_constrained)
+
         loss.backward()
-        optimizer.step()
+
+        def closure():
+            return loss
```
Comment (mohammadbashiri) [inline fitgabor/trainer.py:33]: Shouldn't the whole thing be inside the `closure` function?

``` python
for epoch in pbar:
    def closure():
        optimizer.zero_grad()
        
        # generate gabor
        gabor = gabor_generator()

        # and other stuff
        
        loss = -model_neuron(gabor_std_constrained)
        loss.backward()
        return loss
    
    optimizer.step(closure)
```

ref: https://pytorch.org/docs/stable/optim.html#optimizer-step-closure

**Review by mohammadbashiri — COMMENTED** [review]
Thanks @PPierzc . I left a small comment on the `closure` function.

File: fitgabor/trainer.py (hunk 1 above)
Comment (mohammadbashiri) [reply fitgabor/trainer.py:33]: I actually think this is important because the BFGS optimizer needs to evaluate the whole thing (probably mostly related to gradients) multiple times - so what the closure returns is ideally different every time that it is called by the optimizer. the way it is implemented now the return of closure is always the same.

Comment (PPierzc) [reply fitgabor/trainer.py:33]: Yeah that is why I added it, but I missed the part about it having to return different values.

I am figuring it out now so that it can work correctly with the rest of the code.

Comment (mohammadbashiri) [reply fitgabor/trainer.py:33]: The fact that it returns different values is my assumption (based on intuition), but I have not tried it ever, so I might actually be wrong.

Comment (PPierzc) [reply fitgabor/trainer.py:33]: I implemented this, however, it became really slow (from 5s per neuron to 30s) and NaNs are even more common then previously.

Comment (mohammadbashiri) [reply fitgabor/trainer.py:33]: hm.. ok. 
Could you push the commit? I would also like to take a look (when and if I find time)

Comment (PPierzc) [reply fitgabor/trainer.py]: I pushed the commit

**Comment by mohammadbashiri** [comment]
Already in PR: Closure #6
